text/0000-improve-c-types-for-cross-language-cfi.md

```diff
@@ -0,0 +1,764 @@
+}
+
+unsafe extern "C" fn hello_from_rust(_: c_long) {
+    println!("Hello, world!");
+}
+
+unsafe extern "C" fn hello_from_rust_again(_: c_long) {
+    println!("Hello from Rust again!\n");
+}
+
+fn indirect_call(f: unsafe extern "C" fn(c_long), arg: c_long) {
+    unsafe { f(arg) }
+}
+
+fn main() {
+    // This works
+    indirect_call(hello_from_rust, 1);
+    // This works when using rustc LTO, but does not work when using (proper)
+    // LTO because the Rust compiler and Clang use different encodings for
+    // hello_from_c and the test at the indirect call site at indirect_call.
+    indirect_call(hello_from_c, 2);
+    // This does not work because the Rust compiler and Clang use different
+    // encodings for hello_from_rust_again and the test at the indirect call
+    // site at indirect_call_from_c.
+    unsafe {
+        indirect_call_from_c(hello_from_rust_again, 3);
+    }
+}
+```
+
+example/src/foo.c
+```c
+#include <stdio.h>
+#include <stdlib.h>
+
+void
+hello_from_c(long arg)
+{
+    printf("Hello from C!\n");
+}
+
+void
+indirect_call_from_c(void (*fn)(long), long arg)
+{
+    fn(arg);
+}
+```
+
+Will need to be changed to:
+
+example/src/main.rs
+```rust
+use std::ffi::c_long;
+use std::ffi::cfi;
+
+#[link(name = "foo")]
+extern "C" {
+    fn hello_from_c(_: cfi::c_long);
+    fn indirect_call_from_c(f: unsafe extern "C" fn(cfi::c_long), arg: c_long);
+}
+
+unsafe extern "C" fn hello_from_rust(_: cfi::c_long) {
+    println!("Hello, world!");
+}
+
+unsafe extern "C" fn hello_from_rust_again(_: cfi::c_long) {
+    println!("Hello from Rust again!\n");
+}
+
+fn indirect_call(f: unsafe extern "C" fn(cfi::c_long), arg: c_long) {
+    unsafe { f(cfi::c_long(arg)) }
+}
+
+fn main() {
+    // This will continue to work
+    indirect_call(hello_from_rust, 1);
+    // This will work both when using rustc LTO and when using (proper) LTO
+    // because the Rust compiler and Clang will use the same encoding for
+    // hello_from_c and the test at the indirect call site at indirect_call.
+    indirect_call(hello_from_c, 2);
+    // This will work because the Rust compiler and Clang will use the same
+    // encoding for hello_from_rust_again and the test at the indirect call site
+    // at indirect_call_from_c.
+    unsafe {
+        indirect_call_from_c(hello_from_rust_again, 3);
+    }
+}
+```
+
+example/src/foo.c
+```c
+#include <stdio.h>
+#include <stdlib.h>
+
+void
+hello_from_c(long arg)
+{
+    printf("Hello from C!\n");
+}
+
+void
+indirect_call_from_c(void (*fn)(long), long arg)
+{
+    fn(arg);
+}
+```
+
+Direct calls to extern "C" function types across the FFI boundary, whether CFI
+is enabled or disabled, will continue to work whether Rust integer types or C
+type aliases are used.
+
+For example:
+
+example/src/main.rs
+```rust
+// Optionally, use std::ffi::c_long;
+
+#[link(name = "foo")]
+extern "C" {
+    fn hello_from_c(_: i64);
+    // Or fn hello_from_c(_: c_long);
```

Comment:
You may want to clarify that the `c_long` here does *not* denote `std::ffi::cfi::c_long`. (Otherwise, someone reading the `fn main` below may get the incorrect impression that `1` will automatically be coerced to `cfi::c_long(1)` when necessary.)

Reply:
Makes sense! Done.